File: core/bmc-runtime/src/main/java/org/bmc4j/engine/ContractManifest.java
package org.bmc4j.engine;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

/**
 * The handshake between the {@code bmc-contracts} annotation processor and {@code JbmcBackend}.
 * The processor emits one line per contracted method and one line per generated enforce-proof
 * class into {@value #RESOURCE} on the analysis classpath; the backend reads it back to decide
 * how to rewrite call sites for a given proof:
 *
 * <pre>
 * contract &lt;ownerInternal&gt; &lt;name&gt; &lt;descriptor&gt; &lt;stubOwnerInternal&gt; &lt;stubName&gt;
 * enforce  &lt;proofClassInternal&gt;
 * </pre>
 *
 * <ul>
 *   <li>{@code contract} lines become {@link ContractRewriter.Redirect}s (replace direction).</li>
 *   <li>{@code enforce} lines name the generated proof classes; when one of <em>those</em> is the
 *       analysis entry, the backend excludes it as a caller so the proof sees the real body
 *       (modular enforce). Any other entry is a replace proof and is rewritten fully.</li>
 * </ul>
 */
public final class ContractManifest {

    public static final String RESOURCE = "META-INF/bmc-contracts.txt";

    private final List<ContractRewriter.Redirect> redirects;
    private final Set<String> enforceProofClasses;

    private ContractManifest(List<ContractRewriter.Redirect> redirects, Set<String> enforceProofClasses) {
        this.redirects = redirects;
        this.enforceProofClasses = enforceProofClasses;
    }

    /** Redirects for the replace direction, one per contracted method. */
    public List<ContractRewriter.Redirect> redirects() {
        return redirects;
    }

    /** Internal names of the generated enforce-proof classes. */
    public Set<String> enforceProofClasses() {
        return enforceProofClasses;
    }

    public boolean isEmpty() {
        return redirects.isEmpty() && enforceProofClasses.isEmpty();
    }

    // --- formatting (used by the processor) ---

    public static String contractLine(String ownerInternal, String name, String descriptor,
                                      String stubOwnerInternal, String stubName) {
        return String.join(" ", "contract", ownerInternal, name, descriptor, stubOwnerInternal, stubName);
    }

    public static String enforceLine(String proofClassInternal) {
        return "enforce " + proofClassInternal;
    }

    // --- parsing (used by the backend) ---

    public static ContractManifest parse(List<String> lines) {
        List<ContractRewriter.Redirect> redirects = new ArrayList<>();
        Set<String> enforce = new LinkedHashSet<>();
        for (String raw : lines) {
            String line = raw.trim();
            if (line.isEmpty() || line.startsWith("#")) {
                continue;
            }
            String[] t = line.split("\\s+");
            if (t[0].equals("contract") && t.length == 6) {
                redirects.add(new ContractRewriter.Redirect(t[1], t[2], t[3], t[4], t[5]));
            } else if (t[0].equals("enforce") && t.length == 2) {
                enforce.add(t[1]);
            }
        }
        return new ContractManifest(redirects, enforce);
    }

    /** Read and merge every {@value #RESOURCE} found in the directory entries of {@code classpath}. */
    public static ContractManifest readFromClasspath(String classpath) {
        List<String> lines = new ArrayList<>();
        if (classpath != null) {
            for (String entry : classpath.split(File.pathSeparator)) {
                if (entry.isEmpty()) {
                    continue;
                }
                Path res = Path.of(entry).resolve(RESOURCE);
                if (Files.isRegularFile(res)) {
                    try {
                        lines.addAll(Files.readAllLines(res));
                    } catch (IOException ignored) {
                        // best effort: a manifest we can't read just yields no contracts
                    }
                }
            }
        }
        return parse(lines);
    }
}

File: core/bmc-runtime/src/main/kotlin/org/bmc4j/engine/ContractManifest.kt
package org.bmc4j.engine

import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path

/**
 * The handshake between the `bmc-contracts` annotation processor and `JbmcBackend`.
 * The processor emits one line per contracted method and one line per generated enforce-proof
 * class into [RESOURCE] on the analysis classpath; the backend reads it back to decide
 * how to rewrite call sites for a given proof:
 *
 * ```
 * contract <ownerInternal> <name> <descriptor> <stubOwnerInternal> <stubName>
 * contract <ownerInternal> <name> <descriptor> <stubOwnerInternal> <stubName> instance <stubDescriptor>
 * enforce  <proofClassInternal>
 * ```
 *
 * A 6-token `contract` line is a static target (the call site is `invokestatic`, the stub keeps
 * the descriptor). An 8-token line with `instance` is a pure-instance target: the call site is
 * `invokevirtual`/`invokeinterface`, and `<stubDescriptor>` is the receiver-prepended descriptor
 * of the generated static stub.
 *
 * - `contract` lines become [ContractRewriter.Redirect]s (replace direction).
 * - `enforce` lines name the generated proof classes; when one of *those* is the
 *   analysis entry, the backend excludes it as a caller so the proof sees the real body
 *   (modular enforce). Any other entry is a replace proof and is rewritten fully.
 */
class ContractManifest private constructor(
        private val redirects: List<ContractRewriter.Redirect>,
        private val enforceProofClasses: Set<String>) {

    /** Redirects for the replace direction, one per contracted method. */
    @JvmName("redirects")
    fun redirects(): List<ContractRewriter.Redirect> = redirects

    /** Internal names of the generated enforce-proof classes. */
    @JvmName("enforceProofClasses")
    fun enforceProofClasses(): Set<String> = enforceProofClasses

    val isEmpty: Boolean
        get() = redirects.isEmpty() && enforceProofClasses.isEmpty()

    companion object {

        const val RESOURCE = "META-INF/bmc-contracts.txt"

        // --- formatting (used by the processor) ---

        @JvmStatic
        @JvmOverloads
        fun contractLine(ownerInternal: String, name: String, descriptor: String,
                         stubOwnerInternal: String, stubName: String,
                         instance: Boolean = false, stubDescriptor: String? = null): String {
            val base = listOf("contract", ownerInternal, name, descriptor, stubOwnerInternal, stubName)
            // A static target stays a 6-token line (unchanged format). A pure-instance target appends
            // `instance <stubDescriptor>` so the backend knows to match the virtual call site and emit
            // the receiver-prepended stub descriptor.
            return if (instance) {
                (base + listOf("instance", stubDescriptor ?: descriptor)).joinToString(" ")
            } else {
                base.joinToString(" ")
            }
        }

        @JvmStatic
        fun enforceLine(proofClassInternal: String): String = "enforce $proofClassInternal"

        // --- parsing (used by the backend) ---

        @JvmStatic
        fun parse(lines: List<String>): ContractManifest {
            val redirects = mutableListOf<ContractRewriter.Redirect>()
            val enforce = LinkedHashSet<String>()
            for (raw in lines) {
                val line = raw.trim()
                if (line.isEmpty() || line.startsWith("#")) {
                    continue
                }
                val t = line.split(Regex("\\s+"))
                if (t[0] == "contract" && t.size == 6) {
                    // Static target: invokestatic, stub keeps the descriptor.
                    redirects.add(ContractRewriter.Redirect(t[1], t[2], t[3], t[4], t[5]))
                } else if (t[0] == "contract" && t.size == 8 && t[6] == "instance") {
                    // Pure-instance target: invokevirtual/interface call site, receiver-prepended stub
                    // descriptor in t[7].
                    redirects.add(ContractRewriter.Redirect(t[1], t[2], t[3], t[4], t[5],
                            true, t[7]))
                } else if (t[0] == "enforce" && t.size == 2) {
                    enforce.add(t[1])
                }
            }
            return ContractManifest(redirects, enforce)
        }

        /** Read and merge every [RESOURCE] found in the directory entries of [classpath]. */
        @JvmStatic
        fun readFromClasspath(classpath: String?): ContractManifest {
            val lines = mutableListOf<String>()
            if (classpath != null) {
                for (entry in classpath.split(File.pathSeparator)) {
                    if (entry.isEmpty()) {
                        continue
                    }
                    val res = Path.of(entry).resolve(RESOURCE)
                    if (Files.isRegularFile(res)) {
                        try {
                            lines.addAll(Files.readAllLines(res))
                        } catch (ignored: IOException) {
                            // best effort: a manifest we can't read just yields no contracts
                        }
                    }
                }
            }
            return parse(lines)
        }
    }
}
